src/main/java/com/io/image/manager/controller/ImageController.java

```diff
@@ -17,20 +23,31 @@ public ImageController(ImageService imageService) {
         this.imageService = imageService;
     }
 
-    // TODO: parse query params to image operations and apply them to requested image
     // IMPORTANT: make sure that errors are handled properly
     @ResponseBody
     @RequestMapping(value = "/{filename}", method = RequestMethod.GET, produces = MediaType.IMAGE_JPEG_VALUE)
-    public byte[] getImage(@PathVariable String filename) throws IOException {
+    public byte[] getImage(@PathVariable String filename) throws IOException, ImageOperation.ImageOperationException {
+        List<ImageOperation> operations =  Collections.emptyList();
 
-        // TODO: do the whole image processing
-        // 1. Try to fetch image from cache (stored locally) that has parsed operations already applied
-        // 2. If not found then fetch remote image
-        // 3. If image has been found apply the processing operations
-        // 4. If everything went ok save image to cache
-        // 5. Serve processed image as a response
-        BufferedImage image = imageService.fetchRemoteImage(filename).orElseThrow();
+        Optional<BufferedImage> image;
+        // TODO: do the query params parsing to image operations list
+        image = imageService.fetchLocalImage(filename, operations);
 
+        if (image.isPresent()) {
+            return dumpImage(image.get());
+        }
+
+        image = imageService.fetchRemoteImage(filename);
+        if (image.isPresent()) {
+            BufferedImage img = imageService.applyOperations(image.get(), operations);
```

Comment:
What do you think about move whole logic to the service layer? The controller will have only a few lines and will be look cleaner